raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -339,26 +366,42 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
         }
     }
 
-    private void handleVoteResponse(int remoteNodeId, VoteResponseData response) throws IOException {
-        if (quorum.isLeader()) {
-            logger.debug("Ignoring vote response {} since we already became leader for epoch {}",
+    private boolean handleVoteResponse(
+        RaftResponse.Inbound responseMetadata
+    ) throws IOException {
+        int remoteNodeId = responseMetadata.sourceId();
+        VoteResponseData response = (VoteResponseData) responseMetadata.data;
+        Errors error = Errors.forCode(response.errorCode());
+        OptionalInt responseLeaderId = optionalLeaderId(response.leaderId());
+        int responseEpoch = response.leaderEpoch();
+
+        Optional<Boolean> handled = maybeHandleStandardError(error, responseLeaderId, responseEpoch);
+        if (handled.isPresent()) {
+            return handled.get();
+        } else if (error == Errors.NONE) {
+            if (quorum.isLeader()) {
+                logger.debug("Ignoring vote response {} since we already became leader for epoch {}",
                     response, quorum.epoch());
-        } else if (quorum.isFollower()) {
-            logger.debug("Ignoring vote response {} since we are now a follower for epoch {}",
+            } else if (quorum.isFollower()) {
+                logger.debug("Ignoring vote response {} since we are now a follower for epoch {}",
                     response, quorum.epoch());
-        } else {
-            CandidateState state = quorum.candidateStateOrThrow();
-            if (response.voteGranted()) {
-                state.recordGrantedVote(remoteNodeId);
-                maybeBecomeLeader(state);
             } else {
-                state.recordRejectedVote(remoteNodeId);
-                if (state.isVoteRejected()) {
-                    logger.info("Insufficient remaining votes to become leader (rejected by {}). " +
-                            "We will await expiration of election timeout before retrying",
-                        state.rejectingVoters());
+                CandidateState state = quorum.candidateStateOrThrow();
+                if (response.voteGranted()) {
+                    state.recordGrantedVote(remoteNodeId);
+                    maybeBecomeLeader(state);
+                } else {
+                    state.recordRejectedVote(remoteNodeId);
+                    if (state.isVoteRejected()) {
+                        logger.info("Insufficient remaining votes to become leader (rejected by {}). " +
+                                "We will await expiration of election timeout before retrying",
+                            state.rejectingVoters());
+                    }
                 }
             }
+            return true;
+        } else {
+            return handleUnexpectedError(error, responseMetadata);
```

Comment:
If we can guarantee that if there's any fields in the response that indicate an error that should be handled, then the `error` field is never NONE, then we would never hit this path. Ditto elsewhere.

Reply:
Generally speaking, the 

```
        Optional<Boolean> handled = maybeHandleStandardError(..)
        if (handled.isPresent()) {
            return handled.get();
        } else if (error == Errors.NONE) {
            ...
            return true;
        } else {
            return handleUnexpectedError(error, responseMetadata);
        }
```

pattern can be simplified as

```
if (error == NONE) {
    // ... extra logic
    return true;
} else {
    handled = maybeHandleStandardError(..)  // the error passed in is guaranteed to be not NONE.
    if (handled.isPresent()) {
        return handled.get()
    } else {
        //.. specific handling by the caller, potentially calling handleUnexpectedError
    }
}
```

Personally I feel the above is more straight-forward to understand.

Reply:
I understand where you're coming from, but one of the annoying things is the need to check the epoch in all cases. Even if the error is NONE, we may no longer be interested in the response because we have changed states. I think the nice guarantee that `maybeHandleStandardError` gives you is that the epoch of the response matches the epoch from the request. I found that made it easier to reason about the cases that needed to be handled.